packages/shared/scope-hoisting/src/utils.js

```diff
@@ -67,3 +81,117 @@ export function assertString(v: mixed): string {
   invariant(typeof v === 'string');
   return v;
 }
+
+const DereferenceVisitor = {
+  Identifier(node: Identifier, scope: Scope) {
+    dereferenceIdentifier(node, scope);
+  },
+};
+
+// updates bindings in path.scope.getProgramParent()
+export function pathDereference(path: NodePath<Node>) {
+  walkSimple(path.node, DereferenceVisitor, path.scope.getProgramParent());
+}
+
+// like path.remove(), but updates bindings in path.scope.getProgramParent()
+export function pathRemove(path: NodePath<Node>) {
+  pathDereference(path);
+  path.remove();
+}
+
+function dereferenceIdentifier(node, scope) {
+  let binding = scope.getBinding(node.name);
+  if (binding) {
+    let i = binding.referencePaths.findIndex(v => v.node === node);
+    if (i >= 0) {
+      binding.dereference();
+      binding.referencePaths.splice(i, 1);
+      return;
+    }
+
+    let j = binding.constantViolations.findIndex(v =>
+      Object.values(v.getBindingIdentifiers()).includes(node),
+    );
+    if (j >= 0) {
+      binding.constantViolations.splice(j, 1);
+      if (binding.constantViolations.length == 0) {
+        binding.constant = true;
+      }
+      return;
+    }
+  }
+}
+
+export function removeReplaceBinding(
+  scope: Scope,
+  name: string,
+  newPath: NodePath<
+    | VariableDeclarator
+    | ClassDeclaration
+    | FunctionDeclaration
+    | ImportSpecifier
+    | ImportDefaultSpecifier
+    | ImportNamespaceSpecifier,
+  >,
+  newKind?: VariableDeclarationKind,
+) {
+  let binding = nullthrows(scope.getBinding(name));
+  let path = binding.path;
+  let {node, parent} = path;
+  invariant(
+    isVariableDeclarator(node) && isVariableDeclaration(parent) && !node.init,
+  );
+  // $FlowFixMe Yes this is kind of provide, but ¯\_(ツ)_/¯
+  path._remove();
```

Comment:
Why the _?

Reply:
`path.remove()`ing a declaration also removes the corresponding binding. But we want to keep the binding and only replace the declaration. `path._remove()` merely removes the node in the AST.